{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install dash\n",
    "import dash\n",
    "from dash import dcc, html, dash_table\n",
    "from dash.dependencies import Input, Output\n",
    "import plotly.express as px\n",
    "import pandas as pd\n",
    "import plotly.graph_objects as go\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run Treecluster with different Threshold\n",
    "\n",
    "| Use Case               | Recommended Threshold Range      |\n",
    "|------------------------|----------------------------------|\n",
    "| Outbreak detection     | 0.015–0.03                       |\n",
    "| Species delimitation   | 0.02–0.05                        |\n",
    "| Deep lineage grouping  | 0.05–0.1+                        |\n",
    "\n",
    "\n",
    "To analyze the clusters later on and check the sublineages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Example of how to run the script from the command line:\\npython TreeCluster.py -i /path/to/input/tree.nwk -o /path/to/output/cluster.txt -t 0.01\\n'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Run This Script in Terminal\n",
    "'''Run this script in the terminal with the following command for help:\n",
    "python TreeCluster.py -h\n",
    "'''\n",
    "#!TreeCluster.py -h\n",
    "\n",
    "'''Example of how to run the script from the command line:\n",
    "python TreeCluster.py -i /path/to/input/tree.nwk -o /path/to/output/cluster.txt -t 0.01\n",
    "'''\n",
    "#!TreeCluster.py -i /Users/MiladM-Dev/../tree.nwk -o /Users/MiladM-Dev/../trial0.01.txt -t 0.01"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read files to prepare it for merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_tsv_file = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/metadata.tsv', sep='\\t')\n",
    "\n",
    "#seq_index_tsv_file view\n",
    "#metadata_tsv_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster001 = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/trial0.001.txt', sep='\\t')\n",
    "cluster015 = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/trial0.015.txt', sep='\\t')\n",
    "cluster05 = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/trial0.05.txt', sep='\\t')\n",
    "cluster03 = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/trial0.03.txt', sep='\\t')\n",
    "cluster01 = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/trial0.01.txt', sep='\\t')\n",
    "cluster1 = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/trial0.1.txt', sep='\\t')\n",
    "\n",
    "cluster01= cluster01.rename(columns={'SequenceName': 'ID'})\n",
    "cluster05= cluster05.rename(columns={'SequenceName': 'ID'})\n",
    "cluster03= cluster03.rename(columns={'SequenceName': 'ID'})\n",
    "cluster001= cluster001.rename(columns={'SequenceName': 'ID'})\n",
    "cluster015= cluster015.rename(columns={'SequenceName': 'ID'})\n",
    "cluster1= cluster1.rename(columns={'SequenceName': 'ID'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### describe the clusters data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "714"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''number of clusters'''\n",
    "cluster001['ClusterNumber'].nunique()#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ClusterNumber</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>3463.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>713.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ClusterNumber\n",
       "count         3463.0\n",
       "min             -1.0\n",
       "max            713.0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "cluster001.describe().loc[[\"count\", \"min\", \"max\"]]\n",
    "#cluster1.unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_counts01 = cluster01['ClusterNumber'].value_counts().to_frame(name='Count')\n",
    "cluster_counts01.reset_index(inplace=True)\n",
    "cluster_counts01.columns = ['ClusterNumber01', 'Count']\n",
    "\n",
    "cluster_counts1 = cluster1['ClusterNumber'].value_counts().to_frame(name='Count')\n",
    "cluster_counts1.reset_index(inplace=True)\n",
    "cluster_counts1.columns = ['ClusterNumber1', 'Count']\n",
    "\n",
    "cluster_counts03 = cluster03['ClusterNumber'].value_counts().to_frame(name='Count')\n",
    "cluster_counts03.reset_index(inplace=True)  \n",
    "cluster_counts03.columns = ['ClusterNumber03', 'Count']\n",
    "\n",
    "cluster_counts05 = cluster05['ClusterNumber'].value_counts().to_frame(name='Count')\n",
    "cluster_counts05.reset_index(inplace=True)\n",
    "cluster_counts05.columns = ['ClusterNumber05', 'Count']\n",
    "\n",
    "cluster_counts015 = cluster015['ClusterNumber'].value_counts().to_frame(name='Count')\n",
    "cluster_counts015.reset_index(inplace=True)\n",
    "cluster_counts015.columns = ['ClusterNumber015', 'Count']\n",
    "\n",
    "cluster_counts001 = cluster001['ClusterNumber'].value_counts().to_frame(name='Count')\n",
    "cluster_counts001.reset_index(inplace=True)\n",
    "cluster_counts001.columns = ['ClusterNumber001', 'Count']\n",
    "\n",
    "cluster_counts_combined = pd.concat([cluster_counts001, cluster_counts01, cluster_counts01, cluster_counts015, cluster_counts03, cluster_counts05,  cluster_counts1], axis=1)\n",
    "cluster_counts_combined.to_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/cluster_counts_combined.csv', index=False)  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create metadata CSV include:\n",
    "1. Aligned Sequence\n",
    "2. Location\n",
    "3. Cluster number\n",
    "4. IDs\n",
    "5. years of sampling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add Sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>BLand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00-00022</td>\n",
       "      <td>Nordrhein-Westfalen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00-00035</td>\n",
       "      <td>Nordrhein-Westfalen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>00-00043</td>\n",
       "      <td>Bayern</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>00-00049</td>\n",
       "      <td>Bayern</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>00-00061</td>\n",
       "      <td>Bayern</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3643</th>\n",
       "      <td>24-01146</td>\n",
       "      <td>Nordrhein-Westfalen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3644</th>\n",
       "      <td>24-01149</td>\n",
       "      <td>Nordrhein-Westfalen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3645</th>\n",
       "      <td>24-01150</td>\n",
       "      <td>Hessen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3646</th>\n",
       "      <td>24-01151</td>\n",
       "      <td>Hessen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3647</th>\n",
       "      <td>24-01163</td>\n",
       "      <td>Nordrhein-Westfalen</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3648 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID                BLand\n",
       "0     00-00022  Nordrhein-Westfalen\n",
       "1     00-00035  Nordrhein-Westfalen\n",
       "2     00-00043               Bayern\n",
       "3     00-00049               Bayern\n",
       "4     00-00061               Bayern\n",
       "...        ...                  ...\n",
       "3643  24-01146  Nordrhein-Westfalen\n",
       "3644  24-01149  Nordrhein-Westfalen\n",
       "3645  24-01150               Hessen\n",
       "3646  24-01151               Hessen\n",
       "3647  24-01163  Nordrhein-Westfalen\n",
       "\n",
       "[3648 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"Convert Fasta to DataFrame by Biopython\"\"\"\n",
    "from Bio import SeqIO\n",
    "\n",
    "fasta_file = '/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/aligned.fasta'\n",
    "records = [{\"ID\": record.id, \"seq\": str(record.seq)} for record in SeqIO.parse(fasta_file, \"fasta\")]\n",
    "\n",
    "# Convert to DataFrame\n",
    "df_fasta = pd.DataFrame(records)\n",
    "\n",
    "metadata_tsv_file = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/metadata.tsv', sep='\\t')\n",
    "metadata_tsv_file.rename(columns={'SequenceName': 'ID'}, inplace=True)\n",
    "\n",
    "\"\"\"de_filtered but I lost the last version used for this sequence\"\"\"\n",
    "df_de = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/de_filtered.csv')\n",
    "df_de.rename(columns={'ID_xsl_prefix': 'ID'}, inplace=True)\n",
    "df_de_filtered = df_de[['ID', 'BLand']]\n",
    "df_de_filtered\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#merge\n",
    "\n",
    "merge_1 = pd.merge(cluster01, df_fasta , on='ID')\n",
    "merge_2 = pd.merge(merge_1, metadata_tsv_file, on='ID')\n",
    "merge_3 = pd.merge(merge_2, df_de_filtered, on='ID', how='left')\n",
    "merge_3 = merge_3.rename(columns={'ID': 'SequenceName'})\n",
    "merge_3.to_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/merged_metadata.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add: Genotypes and dates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "markdown"
    }
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (3013603512.py, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[10], line 3\u001b[0;36m\u001b[0m\n\u001b[0;31m    In this section, we will split the DataFrame into subsets based on the genotypes present in the data.\u001b[0m\n\u001b[0m       ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "### Split DataFrame by Genotypes\n",
    "\n",
    "In this section, we will split the DataFrame into subsets based on the genotypes present in the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/c5/srv5gzkd52l_9y41mpknm0t80000gp/T/ipykernel_6561/3495102851.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  B3['SequenceName'] = B3['SequenceName'] + '_B3_' + B3['ClusterNumber'].astype(str)\n",
      "/var/folders/c5/srv5gzkd52l_9y41mpknm0t80000gp/T/ipykernel_6561/3495102851.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  D8['SequenceName'] = D8['SequenceName'] + '_D8_' + D8['ClusterNumber'].astype(str)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SequenceName</th>\n",
       "      <th>ClusterNumber</th>\n",
       "      <th>seq</th>\n",
       "      <th>type</th>\n",
       "      <th>date</th>\n",
       "      <th>BLand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>12-00245_B3_-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2012-06-21</td>\n",
       "      <td>Bayern</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>06-00057_B3_10</td>\n",
       "      <td>10</td>\n",
       "      <td>AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2006-02-20</td>\n",
       "      <td>Baden-Württemberg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>06-00058_B3_10</td>\n",
       "      <td>10</td>\n",
       "      <td>AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2006-02-20</td>\n",
       "      <td>Baden-Württemberg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>07-00450_B3_10</td>\n",
       "      <td>10</td>\n",
       "      <td>AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2007-06-27</td>\n",
       "      <td>Baden-Württemberg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>06-00054_B3_10</td>\n",
       "      <td>10</td>\n",
       "      <td>AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2006-02-17</td>\n",
       "      <td>Baden-Württemberg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2794</th>\n",
       "      <td>23-00184_B3_103</td>\n",
       "      <td>103</td>\n",
       "      <td>---GTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2023-03-20</td>\n",
       "      <td>Bayern</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2795</th>\n",
       "      <td>18-00040_B3_103</td>\n",
       "      <td>103</td>\n",
       "      <td>AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2018-01-10</td>\n",
       "      <td>Baden-Württemberg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2796</th>\n",
       "      <td>11-00756_B3_103</td>\n",
       "      <td>103</td>\n",
       "      <td>AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2011-08-02</td>\n",
       "      <td>Bayern</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2797</th>\n",
       "      <td>11-00742_B3_103</td>\n",
       "      <td>103</td>\n",
       "      <td>AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2011-07-21</td>\n",
       "      <td>Bayern</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2798</th>\n",
       "      <td>11-00718_B3_103</td>\n",
       "      <td>103</td>\n",
       "      <td>AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...</td>\n",
       "      <td>B3</td>\n",
       "      <td>2011-07-07</td>\n",
       "      <td>Bayern</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>464 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SequenceName  ClusterNumber  \\\n",
       "461    12-00245_B3_-1             -1   \n",
       "462    06-00057_B3_10             10   \n",
       "463    06-00058_B3_10             10   \n",
       "464    07-00450_B3_10             10   \n",
       "465    06-00054_B3_10             10   \n",
       "...               ...            ...   \n",
       "2794  23-00184_B3_103            103   \n",
       "2795  18-00040_B3_103            103   \n",
       "2796  11-00756_B3_103            103   \n",
       "2797  11-00742_B3_103            103   \n",
       "2798  11-00718_B3_103            103   \n",
       "\n",
       "                                                    seq type        date  \\\n",
       "461   AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2012-06-21   \n",
       "462   AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2006-02-20   \n",
       "463   AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2006-02-20   \n",
       "464   AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2007-06-27   \n",
       "465   AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2006-02-17   \n",
       "...                                                 ...  ...         ...   \n",
       "2794  ---GTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2023-03-20   \n",
       "2795  AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2018-01-10   \n",
       "2796  AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2011-08-02   \n",
       "2797  AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2011-07-21   \n",
       "2798  AAGGTCAGTTCCACATTGGCATCTGAACTCGGTATCACTGCCGAGG...   B3  2011-07-07   \n",
       "\n",
       "                  BLand  \n",
       "461              Bayern  \n",
       "462   Baden-Württemberg  \n",
       "463   Baden-Württemberg  \n",
       "464   Baden-Württemberg  \n",
       "465   Baden-Württemberg  \n",
       "...                 ...  \n",
       "2794             Bayern  \n",
       "2795  Baden-Württemberg  \n",
       "2796             Bayern  \n",
       "2797             Bayern  \n",
       "2798             Bayern  \n",
       "\n",
       "[464 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "D8 = merge_3[merge_3['type'] == 'D8']\n",
    "B3 = merge_3[merge_3['type'] == 'B3']\n",
    "B3['SequenceName'] = B3['SequenceName'] + '_B3_' + B3['ClusterNumber'].astype(str)\n",
    "D8['SequenceName'] = D8['SequenceName'] + '_D8_' + D8['ClusterNumber'].astype(str)\n",
    "D8.to_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/data/metadata_D8.csv', index=False)\n",
    "B3.to_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/data/metadata_B3.csv', index=False)\n",
    "B3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "134"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "D8 = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/data/metadata_D8.csv')\n",
    "B3 = pd.read_csv('/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/data/metadata_B3.csv')\n",
    "\n",
    "# Filter sequences based on ClusterNumber\n",
    "B3_notclustered = B3[B3['ClusterNumber'] == -1]\n",
    "D8_notclustered = D8[D8['ClusterNumber'] == -1]\n",
    "\n",
    "# Filter sequences based on date range\n",
    "B3_20_24 = B3[(B3['date'] >= '2020-01-01') & (B3['date'] <= '2024-12-31')]\n",
    "D8_20_24 = D8[(D8['date'] >= '2020-01-01') & (D8['date'] <= '2024-12-31')]\n",
    "\n",
    "# Filter sequences based on ClusterNumber, including only those with ClusterNumber > 0, and get representative sequences\n",
    "B3_rep = B3[B3['ClusterNumber'] > 0]\n",
    "B3_rep = B3_rep.drop_duplicates(subset=['ClusterNumber'], keep='first')\n",
    "D8_rep = D8[D8['ClusterNumber'] > 0]\n",
    "D8_rep = D8_rep.drop_duplicates(subset=['ClusterNumber'], keep='first')\n",
    "\n",
    "B3_rep_notclustered = pd.concat([B3_notclustered, B3_rep], ignore_index=True)\n",
    "D8_rep_notclustered = pd.concat([D8_notclustered, D8_rep], ignore_index=True)\n",
    "\n",
    "\n",
    "# Create SeqRecord objects for each sequence in the metadata\n",
    "def create_seq_record(df):\n",
    "    records = []\n",
    "    for _, row in df.iterrows():\n",
    "        seq = Seq(row['seq'])\n",
    "        record = SeqRecord(seq, id=row['SequenceName'], description=\"\")\n",
    "        records.append(record)\n",
    "    return records\n",
    "\n",
    "create_seq_record_B3 = create_seq_record(B3)\n",
    "create_seq_record_B3_20_24 = create_seq_record(B3_20_24)\n",
    "create_seq_record_D8_20_24 = create_seq_record(D8_20_24)\n",
    "create_seq_record_B3_notclustered = create_seq_record(B3_notclustered)\n",
    "create_seq_record_D8_notclustered = create_seq_record(D8_notclustered)\n",
    "create_seq_record_D8 = create_seq_record(D8)\n",
    "create_seq_record_D8_rep = create_seq_record(D8_rep)\n",
    "create_seq_record_B3_rep = create_seq_record(B3_rep)\n",
    "create_seq_record_B3_rep_notclustered = create_seq_record(B3_rep_notclustered)\n",
    "create_seq_record_D8_rep_notclustered = create_seq_record(D8_rep_notclustered)\n",
    "\n",
    "# Write the sequences to a FASTA file\n",
    "output_fasta_file = '/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/data/aligned_'\n",
    "SeqIO.write(create_seq_record_B3, output_fasta_file +'B3.fasta', \"fasta\")\n",
    "SeqIO.write(create_seq_record_D8, output_fasta_file +'D8.fasta', \"fasta\")\n",
    "SeqIO.write(create_seq_record_B3_20_24, output_fasta_file +'B3_20_24.fasta', \"fasta\")\n",
    "SeqIO.write(create_seq_record_D8_20_24, output_fasta_file +'D8_20_24.fasta', \"fasta\")\n",
    "SeqIO.write(create_seq_record_B3_notclustered, output_fasta_file +'B3_notclustered.fasta', \"fasta\")\n",
    "SeqIO.write(create_seq_record_D8_notclustered, output_fasta_file +'D8_notclustered.fasta', \"fasta\")\n",
    "SeqIO.write(create_seq_record_D8_rep, output_fasta_file +'D8_rep.fasta', \"fasta\")\n",
    "SeqIO.write(create_seq_record_B3_rep, output_fasta_file +'B3_rep.fasta', \"fasta\")\n",
    "SeqIO.write(create_seq_record_B3_rep_notclustered, output_fasta_file +'B3_rep_notclustered.fasta', \"fasta\")\n",
    "SeqIO.write(create_seq_record_D8_rep_notclustered, output_fasta_file +'D8_rep_notclustered.fasta', \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster -1:\n",
      "Cluster 10:\n",
      "Cluster 14:\n",
      "Cluster 15:\n",
      "Cluster 19:\n",
      "Cluster 20:\n",
      "Cluster 21:\n",
      "Cluster 22:\n",
      "Cluster 23:\n",
      "Cluster 24:\n",
      "Cluster 25:\n",
      "Cluster 26:\n",
      "Cluster 27:\n",
      "Cluster 40:\n",
      "Cluster 41:\n",
      "Cluster 103:\n",
      "Cluster -1:\n",
      "Cluster 17:\n",
      "Cluster 18:\n",
      "Cluster 35:\n",
      "Cluster 36:\n",
      "Cluster 37:\n",
      "Cluster 38:\n",
      "Cluster 39:\n",
      "Cluster 47:\n",
      "Cluster 48:\n",
      "Cluster 49:\n",
      "Cluster 50:\n",
      "Cluster 51:\n",
      "Cluster 52:\n",
      "Cluster 53:\n",
      "Cluster 54:\n",
      "Cluster 55:\n",
      "Cluster 56:\n",
      "Cluster 57:\n",
      "Cluster 58:\n",
      "Cluster 59:\n",
      "Cluster 72:\n",
      "Cluster 73:\n",
      "Cluster 74:\n",
      "Cluster 75:\n",
      "Cluster 76:\n",
      "Cluster 77:\n",
      "Cluster 78:\n",
      "Cluster 82:\n",
      "Cluster 83:\n",
      "Cluster 84:\n",
      "Cluster 85:\n",
      "Cluster 86:\n",
      "Cluster 87:\n",
      "Cluster 88:\n",
      "Cluster 89:\n",
      "Cluster 90:\n",
      "Cluster 91:\n",
      "Cluster 92:\n",
      "Cluster 93:\n",
      "Cluster 96:\n",
      "Cluster 97:\n",
      "Cluster 98:\n",
      "Cluster 99:\n",
      "Cluster 100:\n",
      "Cluster 101:\n",
      "Cluster 102:\n",
      "Cluster 104:\n",
      "Cluster 105:\n",
      "Cluster 106:\n",
      "Cluster 109:\n",
      "Cluster 112:\n",
      "Cluster 113:\n",
      "Cluster 114:\n",
      "Cluster 115:\n",
      "Cluster 116:\n",
      "Cluster 117:\n",
      "Cluster 118:\n",
      "Cluster 119:\n",
      "Cluster 120:\n",
      "Cluster 121:\n",
      "Cluster 122:\n",
      "Cluster 123:\n",
      "Cluster 124:\n",
      "Cluster 125:\n",
      "Cluster 126:\n",
      "Cluster 127:\n",
      "Cluster 128:\n",
      "Cluster 129:\n",
      "Cluster 130:\n",
      "Cluster 131:\n",
      "Cluster 132:\n",
      "Cluster 133:\n",
      "Cluster 134:\n",
      "Cluster 135:\n",
      "Cluster 136:\n",
      "Cluster 137:\n",
      "Cluster 138:\n",
      "Cluster 139:\n",
      "Cluster 140:\n",
      "Cluster 141:\n",
      "Cluster 142:\n"
     ]
    }
   ],
   "source": [
    "# Split into a dictionary of DataFrames keyed by cluster number\n",
    "B3_by_cluster = {cluster: group for cluster, group in B3.groupby('ClusterNumber')}\n",
    "D8_by_cluster = {cluster: group for cluster, group in D8.groupby('ClusterNumber')}\n",
    "# For example, DataFrame for cluster 0\n",
    "#df_cluster_0 = dfs_by_cluster[0]\n",
    "\n",
    "# To view all:\n",
    "for cluster_id, df_cluster in B3_by_cluster.items():\n",
    "    print(f\"Cluster {cluster_id}:\")\n",
    "    create_seq_record(df_cluster)\n",
    "    df_cluster.to_csv(f'/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/data/B3_cluster_{cluster_id}.csv', index=False)\n",
    "    SeqIO.write(create_seq_record(df_cluster), f'/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/data/B3_cluster_{cluster_id}.fasta', \"fasta\")\n",
    "\n",
    "for cluster_id, df_cluster in D8_by_cluster.items():\n",
    "    print(f\"Cluster {cluster_id}:\")\n",
    "    create_seq_record(df_cluster)\n",
    "    df_cluster.to_csv(f'/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/data/D8_cluster_{cluster_id}.csv', index=False)\n",
    "    SeqIO.write(create_seq_record(df_cluster), f'/Users/MiladM-Dev/Documents/1PhD/project-1-N450/project-1.1-gendata/Nextstrain/trial-1/results/clusters/data/D8_cluster_{cluster_id}.fasta', \"fasta\")\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
